sendrecv/gst-rust/src/main.rs

```diff
@@ -0,0 +1,480 @@
+            "candidate": candidate,
+            "sdpMLineIndex": mlineindex,
+          }
+        });
+    app_control.ws_sender.send(message.to_string()).unwrap();
+    None
+}
+
+fn start_pipeline(app_control: &Arc<Mutex<AppControl>>) -> Result<gst::Element, Error> {
+    let pipe = gst::parse_launch(&format!(
+        "webrtcbin name=sendrecv {}
+        videotestsrc pattern=ball ! videoconvert ! queue ! vp8enc deadline=1 ! rtpvp8pay !
+        queue ! {}96 ! sendrecv.
+        audiotestsrc wave=red-noise ! audioconvert ! audioresample ! queue ! opusenc ! rtpopuspay !
+        queue ! {}97 ! sendrecv.
+        ",
+        STUN_SERVER, RTP_CAPS_VP8, RTP_CAPS_OPUS
+    ))?;
+
+    let webrtc = pipe.clone()
+        .dynamic_cast::<gst::Bin>()
+        .unwrap()
+        .get_by_name("sendrecv")
+        .unwrap();
+    let app_control_clone = app_control.clone();
+    webrtc.connect("on-negotiation-needed", false, move |values| {
+        on_negotiation_needed(&app_control_clone, values)
+    })?;
+
+    let app_control_clone = app_control.clone();
+    webrtc.connect("on-ice-candidate", false, move |values| {
+        send_ice_candidate_message(&app_control_clone, values)
+    })?;
+
+    let pipe_clone = pipe.clone();
+    // TODO: replace with webrtc.connect_pad_added
+    webrtc.connect("pad-added", false, move |values| {
+        on_incoming_stream(values, &pipe_clone)
+    })?;
+
+    pipe.set_state(gst::State::Playing).into_result()?;
+
+    return Ok(webrtc);
+}
+
+struct WsClient {
+    webrtc: Option<gst::Element>,
+    app_control: Arc<Mutex<AppControl>>,
+}
+
+struct AppControl {
+    app_state: AppState,
+    ws_sender: ws::Sender,
+    peer_id: String,
+}
+
+impl WsClient {
+    fn update_state(&self, state: AppState) {
+        self.app_control.lock().unwrap().app_state = state
+    }
+}
+
+impl ws::Handler for WsClient {
+    fn on_open(&mut self, _: ws::Handshake) -> ws::Result<()> {
+        self.update_state(AppState::ServerConnected);
+        self.update_state(register_with_server(&self.app_control.clone()));
+        Ok(())
+    }
+
+    fn on_message(&mut self, msg: ws::Message) -> ws::Result<()> {
+        // Close the connection when we get a response from the server
+        let msg_text = msg.into_text().unwrap();
+        if msg_text == "HELLO" {
+            if self.app_control.lock().unwrap().app_state != AppState::ServerRegistering {
+                panic!("ERROR: Received HELLO when not registering");
+            }
+            self.update_state(AppState::ServerRegistered);
+            setup_call(&self.app_control.clone());
+            return Ok(());
+        }
+        if msg_text == "SESSION_OK" {
+            if self.app_control.lock().unwrap().app_state != AppState::PeerConnecting {
+                panic!("ERROR: Received SESSION_OK when not calling");
+            }
+            self.update_state(AppState::PeerConnected);
+            self.webrtc = match start_pipeline(&self.app_control) {
```

Comment:
The assertion, `update_state` and `start_pipeline` seem like something that could all be moved into a single function inside a `impl AppControl` block (after refactoring as mentioned above). That way you have one function for each message, the assertions locally in there, and here you only dispatch based on the message type